src/basic/conf-virt.c

```diff
@@ -0,0 +1,290 @@
+                }
+        }
+
+        return false;
+}
+
+static ConfidentialVirtualization
+detect_sev(void)
+{
+        uint32_t eax, ebx, ecx, edx;
+        uint64_t msrval;
+
+        eax = CPUID_GET_HIGHEST_FUNCTION;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+
+        if (eax < CPUID_AMD_GET_ENCRYPTED_MEMORY_CAPABILITIES)
+                return CONFIDENTIAL_VIRTUALIZATION_NONE;
+
+        eax = CPUID_AMD_GET_ENCRYPTED_MEMORY_CAPABILITIES;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+
+        /* bit 1 == CPU supports SEV feature
+         *
+         * Note, Azure blocks this CPUID leaf from its SEV-SNP
+         * guests, so we must fallback to probing the TPM which
+         * exposes a SEV-SNP attestation report as evidence.
+         */
+        if (!(eax & (1 << 1))) {
+                log_debug("No sev in CPUID, try hyperv CPUID");
+
+                if (detect_hyperv_sev()) {
+                        return CONFIDENTIAL_VIRTUALIZATION_SEV_SNP;
+                } else {
+                        log_debug("No hyperv CPUID");
+                }
+                return CONFIDENTIAL_VIRTUALIZATION_NONE;
+        }
+
+        msrval = msr(MSR_AMD64_SEV);
+
+        /* Test reverse order, since the SEV-SNP bit implies
+         * the SEV-ES bit, which implies the SEV bit */
+        if (msrval & (1 << 2)) {
+                return CONFIDENTIAL_VIRTUALIZATION_SEV_SNP;
+        } else if (msrval & (1 << 1)) {
+                return CONFIDENTIAL_VIRTUALIZATION_SEV_ES;
+        } else if (msrval & (1 << 0)) {
+                return CONFIDENTIAL_VIRTUALIZATION_SEV;
+        }
+        return CONFIDENTIAL_VIRTUALIZATION_NONE;
+}
+
+static ConfidentialVirtualization detect_tdx(void)
+{
+        uint32_t eax, ebx, ecx, edx;
+        char sig[13];
+
+        eax = CPUID_GET_HIGHEST_FUNCTION;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+        log_debug("CPUID max function: %x %x %x %x", eax, ebx, ecx,edx);
+
+        if (eax < CPUID_INTEL_TDX_ENUMERATION)
+                return CONFIDENTIAL_VIRTUALIZATION_NONE;
+
+        memset(sig, 0, sizeof sig);
+        cpuid_leaf(CPUID_INTEL_TDX_ENUMERATION, sig, true);
+
+        if (memcmp(sig, CPUID_SIG_INTEL_TDX, sizeof(sig)) == 0)
+                return CONFIDENTIAL_VIRTUALIZATION_TDX;
+
+        return CONFIDENTIAL_VIRTUALIZATION_NONE;
+}
+
+static bool
+detect_hypervisor(void)
+{
+        uint32_t eax, ebx, ecx, edx;
+        bool is_hv;
+
+        eax = CPUID_PROCESSOR_INFO_AND_FEATURE_BITS;
+        ebx = ecx = edx = 0;
+
+        cpuid(&eax, &ebx, &ecx, &edx);
+
+        is_hv = ecx & CPUID_FEATURE_HYPERVISOR;
+
+        log_debug("CPUID is hypervisor: %s", is_hv ? "yes" : "no");
+        return is_hv;
+}
+
+
+ConfidentialVirtualization detect_confidential_virtualization(void) {
+        char sig[13];
+
+        /* Skip everything on bare metal */
+        if (!detect_hypervisor())
```

Comment:
the existing condition_virtualization stuff should work for this I think? Rather than defining a new detect_hypervisor()

Reply:
I very specifically wanted to check the 'hypervisor' feature bit in ECX returns from CPUID leaf 0x1, because that is part of the documented sequence of CPUID checks for detecting TDX from Intel.

AMD didn't document a specific sequence, but checking the 'hypervisor' feature bit makes sense there too, so did it upfront here.  

The existing `detect_vm()` check would be too weak, for this IMHO, as I don't believe it guarantees this.

Reply:
ok, let's keep it this way for now then